src/Storages/MergeTree/MergeTreeBaseSelectProcessor.cpp

```diff
@@ -67,6 +67,24 @@ Chunk MergeTreeBaseSelectProcessor::generate()
     return {};
 }
 
+/* setLimits caches once passed leaf limits and ignores subsequent changes.
+ * This protects leaf limits from limit overrides on different query stages.
```

Comment:
This is the most suspicious place, I don't see how it can be, i.e. called multiple times for the same table in the query (maybe you mixed up remote query and the local?)

Reply:
Right. Here's what I observe. On testing stand I have two CH servers forming the cluster of 2 shards. I have a local tables on every shard and distributed table to get the `global view`. Then I send the following query to server-1:
```
SELECT count() FROM (SELECT * FROM test_distributed) SETTINGS max_rows_to_read_leaf = 60000;
```
On server-2 I see the trace logs saying that query is accepted and https://github.com/ClickHouse/ClickHouse/blob/7524a6d412f01ed455dcc36f1ba334f5ba4e332c/src/Processors/QueryPlan/ReadFromStorageStep.cpp#L119 applied successfully. 
Server-1 on its turn should do the following:
1. Apply the leaf limits to processor that reads from the local table;
2. Ignore the leaf limits for processor that merges results from local and remote tables.

So the function `ReadFromStorageStep` on server-1 is called twice while building the pipeline and processors, so the limits are applied [two times as well ](https://github.com/ClickHouse/ClickHouse/blob/7524a6d412f01ed455dcc36f1ba334f5ba4e332c/src/Processors/QueryPlan/ReadFromStorageStep.cpp#L129). But the way how [setLimits implemented](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Pipe.cpp#L782) overrides limits for all processors neglecting previously set leaf limit. And as result, leaf limits won't be applied on root node while reading from local table, which is not correct.
That's why I'm caching the first attempt to set the leaf limits and ignore subsequent attempts. Means, once the limit was set - it shouldn't be dropped after that. My understanding of the process may be incorrect or not full, but proved by observations.

Reply:
MergeTreeBaseSelectProcessor::setLimits will be called by the number of parts and later for distributed query too.
The code is really tricky (and at least requires good comment), maybe it will be cleaner if there will be separate SourceWithProgress::setLeafLimits, although not sure.

AFAIR @KochetovNicolai was planning IStorage::read returns IQueryPlanStep, I guess this will solve the problem (since the limit can be set in `SelectStreamFactory::createForShard`, @KochetovNicolai can you comment on this?

Reply:
There is a draft already #14815

Reply:
Indeed.
`setLimits` sets limits for all sources which are created for query, which include distributed source.
I am going to finish #14815 soon